juju/juju
# uniter: Smarter config-changed pt3 - address config

## Description of change

This is a follow-up to #9498 and #9501 to prevent config-changed from being run unless the config has actually, like, changed.

Add Machine.WatchAddressesHash and Unit.WatchContainerAddressesHash that emit the hash of the current addresses, and feed that down to the uniter (in the RemoteStateWatcher).

This is the last item that can trigger config-changed so I've removed the ConfigVersion field from local state and the remote state snapshot. That means that we only fire the hook when one or more of the config, trust and address hashes is different from the state on disk. There were a lot of uniter tests that assumed that config-changed would always run after the unit agent restarted, those have been fixed - in `UniterSuite.TestUniterRelations` there was one that was using the hook to extract relation data so I change configuration explicitly so that still works.

`UniterSuite.TestRunCommand` has a workaround fix until I can come back to sort out what the underlying problem is.

## QA steps

* Deploy an application to a new model.
* See that the config-changed hook for the new unit runs after install.
* Look in the uniter state file to see the hashes for config, trust and addresses.
* Change the config, see that the config-changed hook runs and the hash in the state file is updated.
* Do the same for trust.
* Not sure how to change addresses for the unit, but if you can, do that too.
* Stop and start the unit agent.
* Config-changed shouldn't fire.
* Stop and start the controller.
* When the unit agent reconnects it shouldn't run config-changed.

## Documentation changes

We will need to clearly highlight this behaviour change in charming documentation - it's likely there are some charms that (incorrectly) rely on the config-changed hook running after agent start to start their workload, rather than setting up service files.
